---
title: "Phenotyping of gastric clones from RNA-seq data -  HISAT2 alignment to hg38, featureCounts"
author: "Hilmar Berger"
output: 
 html_document:
    toc: true
    toc_depth: 3
    toc_float: false
    number_sections: true
    code_folding: hide
    
pdf_document:
    fig_caption: true
date: "`r format(Sys.time(), '%d %B, %Y %H:%M:%S')`"
---

```{r, warning=FALSE, message=FALSE}
rm(list=ls())

library(readxl)
library(reshape2)
library(pheatmap)
library(ggplot2)
library(knitr)
library(scales)
library(data.table)
library(DESeq2)

result_folder = paste("./Results",format(Sys.time(), "%Y-%m-%d"),sep="/")
if (!file.exists(result_folder)) dir.create(result_folder, recursive=T)

```

# Input files

The following input files are required: 

  - Feature counts output file as generated by featureCounts. All samples must be contained in one file. 
  - Gene annotation file (Ensembl_v98_human_gene_annotation.txt)
  - Sample description file, containing at least columns *sampleID*, *Patient* and *Condition*. *sampleID* must be a unique identifier of the sample that matches the one in the featureCounts input file (suffixes in the count table can be removed by setting *file_suffix* in the R code below). *Patient* should identify unique patients, and *Condition* should describe the sample class (for example IM, non-IM, intestine-control or others).
  - A list of genes up- or downregulated in Colon compared to Stomach (*Colon_vs_Stomach_DGE.txt*)
  

```{r}
# ADAPT THIS TO YOUR FILES
colon_vs_stomach_signature_file = "Colon_vs_Stomach_DGE.txt"
  
```


```{r}
IMPORT_DATA = FALSE

if(IMPORT_DATA) {
  suppressMessages(library(org.Hs.eg.db))
  suppressMessages(library(AnnotationDbi))

  count_data_hisat = fread("../featureCount/ESGIv2_China_counts.txt")
  sample_cols = grepl(".bam$", colnames(count_data_hisat), perl=T)

  count_mat = as.matrix(count_data_hisat[, sample_cols, with=F])
  rownames(count_mat) = count_data_hisat$Geneid
  snames = gsub("_sorted\\.bam", "", unlist(sapply(strsplit(colnames(count_mat), "/"), function(x) x[length(x)] )) )
  colnames(count_mat) = snames
  
  gene_anno_a = as.data.frame(count_data_hisat[, !sample_cols, with=F])
  
  # library(biomaRt)
  # ensembl90 <- useEnsembl(biomart = 'ENSEMBL_MART_ENSEMBL', dataset = 'hsapiens_gene_ensembl', version = 90)
  # gg = getBM(attributes = c("entrezgene","hgnc_symbol","ensembl_gene_id", "description"), filter = c("chromosome_name", "ensembl_gene_id") , 
  #          values=list(chromosome_name = c(as.character(1:22), "X","Y"), ensembl_gene_id = gene_anno$Geneid), mart=ensembl90 )
  # fwrite(gg, file="Ensembl_v90_human_gene_annotation.txt", sep="\t", quote=F)
  # 
  # ens_anno_orig = fread("Ensembl_v90_human_gene_annotation.txt")
  # setnames(ens_anno_orig, colnames(ens_anno_orig), c("EntrezID","GeneSymbol", "Geneid", "GeneName"))
  # # as usual there are 1:n relationships to GeneSymbol, EntrezID, GeneName - fix those to 1:1
  # collapse_fun <- function(x) paste(unique(x[!is.na(x)]), collapse="," )
  # ens_anno_agg = ens_anno_orig[, .(EntrezID=collapse_fun(EntrezID), GeneSymbol=collapse_fun(GeneSymbol), GeneName=collapse_fun(GeneName)), by = "Geneid" ]

  tx_anno = read.table("/home/hilmar/Work/References/Human/Annotations/hg38/Gencode/v35/gencode.v35.transcript.anno.txt", sep="\t", header=T, stringsAsFactors = F)

  entrez_from_gencode =  read.table(gzfile("/home/hilmar/Work/References/Human/Annotations/hg38/Gencode/v35/gencode.v35.metadata.EntrezGene.gz"), sep="\t", header=T, stringsAsFactors = F)
  colnames(entrez_from_gencode) = c("transcript","EntrezID")
  # There is a number (n=27) transcripts in this Gencode version that have more than one Entrez ID, from two clusters of EntrezIDs pointing to a LOC... gene. Since we cannot decide which of the EntrezID is the best one we here just remove all Entrez IDs from those transcripts
  dup_tx = unique(entrez_from_gencode[which(duplicated(entrez_from_gencode$transcript)),"transcript"])
  entrez_from_gencode = subset(entrez_from_gencode, !transcript %in% dup_tx)
  tx_anno = merge(tx_anno, unique(entrez_from_gencode), by="transcript", all.x=T, sort=F)
  rownames(tx_anno) = tx_anno$transcript

  gene_anno = unique(tx_anno[,c("gene","GeneSymbol", "EntrezID")])
  # for some of the genes there are transcripts with a valid EntrezID and others without it (NA). We here remove the NA entries in order to make the mapping unique. 
  tmp = as.data.table(gene_anno)[, .(NAentry = any(is.na(EntrezID)), count=.N), by="gene"]
  genes_with_additional_NA = xx = subset(tmp, NAentry & count > 1)$gene
  gene_anno = subset(gene_anno, !(gene %in% genes_with_additional_NA & is.na(EntrezID)) )
  gene_anno$gene_fixed = unlist(sapply(strsplit(gene_anno$gene,"\\."), function(x) x[1]))
  
  genename=select(org.Hs.eg.db, keys=gene_anno$gene_fixed,         columns=c("GENENAME"),keytype="ENSEMBL") 
  GeneName = as.data.frame(tapply(genename$GENENAME, genename$ENSEMBL, paste, collapse=","))
  colnames(GeneName) = "GeneName"
  gene_anno = merge(gene_anno, GeneName, by.x="gene_fixed", by.y=0, sort=F, all.x=T)
  rownames(gene_anno) = gene_anno$gene
  
  gene_anno = merge(gene_anno_a, gene_anno, by.x="Geneid", by.y="gene", all.x=T, sort=F)
  rownames(gene_anno) = gene_anno$Geneid
  
  # FIXME consider reading total counts (including unmapped/unassigned) for normalization
  fpkm_mat = sweep(sweep(count_mat, 2, apply(count_mat, 2, sum)/1e6, "/"), 1, gene_anno[rownames(count_mat), ]$Length/1e3, "/")
  
  save(count_mat, fpkm_mat, gene_anno, file="HISAT_featureCount_hg38_GencodeV35_gene_expression.Rdata")
} else {
  load("HISAT_featureCount_hg38_GencodeV35_gene_expression.Rdata")
}

```

```{r}
exp_design = as.data.frame(read_xlsx("../../metadata/phenotyping/SampleDescription.xlsx", sheet = 1), stringsAsFactors=FALSE)
exp_design$Group = ifelse(grepl("^NON IM",exp_design$Remark), "Non-IM", ifelse(grepl("^IM", exp_design$Remark),"IM", "Control"))
exp_design$Patient = gsub("-|\\s","", gsub("^IM|^NON IM","", gsub("A$|AC$|C$","", exp_design$Remark)))

sra_info = fread("../../metadata/phenotyping/SraRunInfo.csv", sep=",")

exp_design = merge(exp_design, sra_info[, c("Run","SampleName","BioSample","spots")], by.x="Name in NCBI", by.y="SampleName", all.x=T, sort=F)
colnames(exp_design) = gsub("Run","sampleID", colnames(exp_design))
rownames(exp_design) = exp_design$sampleID
ed = exp_design

ed$Condition = ed$Group

```


```{r, message=FALSE, warning=FALSE}
coldata <- ed[colnames(count_mat),]
rownames(coldata) <- coldata$sampleID

dds = DESeqDataSetFromMatrix(count_mat, colData = coldata, design = ~ Condition )

tmp = as.data.frame(assay(vst(dds, blind=F)))
norm_counts = merge(tmp, gene_anno[, c("Geneid","EntrezID","GeneSymbol","GeneName")], by.x=0, by.y="Geneid", all.x=T, sort=F)
colnames(norm_counts) = gsub("Row.names", "Geneid", colnames(norm_counts))
```

# Data overview

We first check the expression of a number of several known intestine associated genes across all samples

## Known IM/intestinal markers

```{r, fig.width=10, fig.height=6}
sel_genes = c("GKN1","GKN2","PGC","MUC5AC","MUC6","TFF3","VIL1","CDX1","CDX2","ISX", "MUC2", "SPINK4","ITLN1","TFF2","DEFA5", "TFF1", "REG4", "LRIG1", "OLFM4")

sel_samples = ed$sampleID
nc_selected = subset(norm_counts, GeneSymbol %in% sel_genes )
nc_mat_selected = as.matrix(nc_selected[,sel_samples])
nc_mat_selected = nc_mat_selected + rnorm(ncol(nc_mat_selected) * nrow(nc_mat_selected), 0,0.01)
rownames(nc_mat_selected) = nc_selected$GeneSymbol
col_anno = ed[, c("Condition","Patient")]
rownames(col_anno) = ed$sampleID
breaks_new=c(-10,-5,-3,seq(-2,2,by=4/94),3,5,10)
pheatmap(nc_mat_selected, scale="row", main="Known IM markers" , annotation_col = col_anno, breaks=breaks_new, labels_col = ed[colnames(nc_mat_selected), "label_full"])

```

## Top Intestinal/Goblet Markers

```{r, fig.width=10, fig.height=6}
sel_genes = c("VIL1","CDX2","SPINK4", "REG4", "MUC2")

sel_samples = ed$sampleID
nc_selected = subset(norm_counts, GeneSymbol %in% sel_genes )
nc_mat_selected = as.matrix(nc_selected[,sel_samples])
nc_mat_selected = nc_mat_selected + rnorm(ncol(nc_mat_selected) * nrow(nc_mat_selected), 0,0.01)
rownames(nc_mat_selected) = nc_selected$GeneSymbol
col_anno = ed[, c("Condition","Patient")]
rownames(col_anno) = ed$sampleID
breaks_new=c(-10,-5,-3,seq(-2,2,by=4/94),3,5,10)
pheatmap(nc_mat_selected, scale="row", main="Known IM markers" , annotation_col = col_anno, breaks=breaks_new, labels_col = ed[gsub("^X", "", colnames(nc_mat_selected)), "Library.Name"])

```


# Selected genes associated with Colon vs. Stomach

Here we select only the genes determined as differentially expressed between 
Colon and Stomach (generated using data from colon biobank normal tissues [van Wetering et al., PMID: 25957691] and gastric glands [Bartfeld et al ,PMID: 25307862] )

```{r}
# Colon vs. Stomach signature
# colon_sig_tmp = read.table("/home/hilmar/Work/DataSets/MPIIB/data_genome2/public_data/GSE64392_colon_biobank/analysis/Results/2016-08-22/Differential_expression_results_colon_vs_stomach_glands.txt", sep="\t", header=T, stringsAsFactors = F)
# colon_vs_stomach_up = subset(colon_sig_tmp, logFC > 1 & adj.P.Val < 0.05)
# colon_vs_stomach_down = subset(colon_sig_tmp, logFC < -1 & adj.P.Val < 0.05)
# 
# colon_vs_stomach_dge = rbind(colon_vs_stomach_up, colon_vs_stomach_down)
# fwrite(colon_vs_stomach_dge, "Colon_vs_Stomach_DGE.txt", sep="\t", quote=F)

colon_vs_stomach_dge = fread(colon_vs_stomach_signature_file)

colon_vs_stomach_up_and_down = unique(data.frame(GeneSymbol = colon_vs_stomach_dge$SYMBOL, group = ifelse(colon_vs_stomach_dge$logFC>0, "Colon","Stomach"), stringsAsFactors = F))
rownames(colon_vs_stomach_up_and_down) = colon_vs_stomach_up_and_down$GeneSymbol

```


## Multi Dimensional Scaling on Colon and Stomach specific Genes

We would like to know how samples are located within the expression space defined by colon or stomach associated genes. Note that this is strongly driven by the variance structure of the data, so even though the genes are restricted to the ones relevant for the colon-stomach distinction, the positioning of individual samples can be biased towards the latent factors in the current data set (e.g. batch effects or similar). 

In an ideal case we would expect IM cases (and possible intestinal controls like small intestine, duodenum, colon) on one side of the 1st coordinated and non-IM as well as optional healthy stomach controls on the opposite side. 

```{r, fig.width=8, fig.height=8}
sel_genes = unique(colon_vs_stomach_up_and_down$GeneSymbol)

cp = palette(rainbow(length(unique(ed$Condition))))
sel_samples = ed$sampleID
tmp = subset(norm_counts, GeneSymbol %in% sel_genes )
tmp2 = as.matrix(tmp[,sel_samples])
rownames(tmp2) = tmp$GeneSymbol
data_inp = t(log10(tmp2 +1))
rownames(data_inp) = rownames(data_inp)

d <- dist(data_inp) # euclidean distances between the rows
fit <- cmdscale(d,eig=TRUE, k=2) # k is the number of dim

# plot solution
x <- fit$points[,1]
y <- fit$points[,2]
gg = factor(ed[rownames(data_inp),]$Condition)
plot(x, y, xlab="Coordinate 1", ylab="Coordinate 2", main="Metric MDS, all samples", type="p", pch=20, xlim=1.25*c(min(x),max(x)))
text(x,y,labels=ed[rownames(data_inp),]$label, col=cp[as.numeric(gg)], adj = c(-0.05,0.5))
legend("topleft", legend=levels(gg), fill=cp[1:length(levels(gg))])
abline(v=0, lwd=2, col="brown")
```


## PCA

Here we run PCA (similar to MDS above) on all samples. The main goal is to derive weights for individual genes associated with the main latent factors in the data set (i.e. those associated with the first principal component) and to derive some (possibly biased) score for each sample.

```{r, fig.width=8, fig.height=8}
## Primary Component Analysis on normalized data after adjusting for patient effect
norm_exp = data_inp
NA_cols = apply(norm_exp,2,function(x) sum(is.na(x)))
pca = prcomp(norm_exp[,NA_cols==0])
#pca = prcomp(t(normalized$E), na.action=na.omit)
plot(pca$x[,1],pca$x[,2],type="p", xlab="1st principal component",ylab="2nd principal component", main="PCA on normalized expression data", pch=20)
gg = as.factor(ed[rownames(norm_exp),]$Condition)
cp = palette(rainbow(length(levels(gg))))

text(pca$x[,1],pca$x[,2],labels=ed[rownames(norm_exp),]$label, col=cp[as.numeric(gg)], cex=0.7, adj =-0.1)
abline(h=0, v=0)
legend("topleft", legend=levels(gg), fill=cp[1:length(levels(gg))])
```

```{r, fig.width=16}
weights = pca$rotation[, 1:3]
weights = weights[order(weights[, "PC1"]),]
```

### Genes associated with main latent factor (i.e. 1st principal component)

Note that the signs associated with the 1st PC are arbitrary. One group should contain at least a few intestine associated marker genes (e.g. VIL1, REG4 or others) while the other one should contain at least a few stomach associated marker genes (e.g. IRX2, MUC6 or others).


#### Genes most associated with first class


```{r}
head(weights,20)
```

#### Genes most associated with second class

```{r}
tail(weights, 20)
```

#### 1st PC per sample

If there are no other strong latent factors in the data set, the first principal component should describe the distinction between intestinal-like and stomach-like samples, i.e. one group should have positive (or negative) scores and the other group should show the opposite sign. 

```{r, fig.width=14, fig.height = 8 }
tmp = pca$x[,1]
tmp = sort(tmp)
gg = as.factor(ed[names(tmp), "Condition"])
par(mar=c(8,4,4,1))
barplot(tmp, las=2, col=as.numeric(gg), ylab="1st PC", names.arg = ed[names(tmp),"labels_full"])
legend("topleft", legend=levels(gg), fill=cp[1:length(levels(gg))])
```

#### Heatmap of top genes associated with IM or non-IM

```{r, fig.width=14, fig.height=8}
top_colon = rownames(head(weights,20))
bottom_colon = rownames(tail(weights,20))

sel_genes = c(top_colon, bottom_colon, "CDX2","MUC2")

sel_samples = ed$sampleID
nc_selected = subset(norm_counts, GeneSymbol %in% sel_genes )
nc_mat_selected = as.matrix(nc_selected[,sel_samples])
nc_mat_selected = nc_mat_selected + rnorm(ncol(nc_mat_selected) * nrow(nc_mat_selected), 0,0.01)
rownames(nc_mat_selected) = nc_selected$GeneSymbol

col_anno = ed[, c("Condition","Patient")]
rownames(col_anno) = ed$sampleID
breaks_new=c(-10,-5,-3,seq(-2,2,by=4/94),3,5,10)

row_anno = colon_vs_stomach_up_and_down[, "group", drop=F]

pheatmap(nc_mat_selected, scale="row", main="Top20 associated with IM and non-IM" , annotation_col = col_anno, breaks=breaks_new, labels_col = ed[gsub("^X", "", colnames(nc_mat_selected)), "Library.Name"],  annotation_row = row_anno)

```


# Unbiased score computed using all genes up or down in Colon vs. Stomach

We here compute the median expression of all genes up or down-regulated in Colon vs. Stomach, respectively, for each sample. We then normalize those median values per sample by substracting the global median across all samples for each of the two parameters. 

Finally, we substract the normalized stomach score from the normalized colon score to obtain the final value. 


```{r, fig.width=14, fig.height=8}

sel_genes = unique(colon_vs_stomach_up_and_down$GeneSymbol)

sel_samples = ed$sampleID
nc_selected = subset(norm_counts, GeneSymbol %in% sel_genes )
nc_mat_selected = as.matrix(nc_selected[,sel_samples])
nc_mat_selected = nc_mat_selected + rnorm(ncol(nc_mat_selected) * nrow(nc_mat_selected), 0,0.01)
rownames(nc_mat_selected) = nc_selected$GeneSymbol


tmp = split(as.data.frame(nc_mat_selected), factor(colon_vs_stomach_up_and_down[rownames(nc_mat_selected), "group"]))
medians_per_sample_and_group = do.call(rbind, lapply(tmp, apply, 2, median))
global_means = apply(medians_per_sample_and_group, 1, median)

scores_normalized = sweep(medians_per_sample_and_group, 1, global_means, "-")
col_anno = ed[, c("Condition","Patient")]
rownames(col_anno) = ed$sampleID
pheatmap(scores_normalized, annotation_col = col_anno, labels_col = ed[gsub("^X", "", colnames(scores_normalized)), "label_full"])


scores_final = scores_normalized["Colon",]-scores_normalized["Stomach",]
scores_final = scores_final[order(-scores_final)]
par(mar=c(14,4,4,1))
barplot(scores_final, col = as.numeric(as.factor(col_anno[names(scores_final),"Condition"])), las=2, names.arg = ed[gsub("^X", "", names(scores_final)), "Remark"], ylab="Colon-vs-Stomach score")

```

Final scores are written to file *Score_table.txt*

```{r}
tmp = as.data.frame(t(scores_normalized))
tmp$Score = tmp$Colon-tmp$Stomach
tmp$Label = ed[rownames(tmp), "label"]
tmp$sampleID = rownames(tmp)

score_table = tmp[order(-tmp$Score),]

fwrite(score_table, file="Score_table.txt", sep="\t", quote=F)
```

