---
title: "Stomach Methylation - Preprocessing and QC"
author: "Hilmar Berger, based on code by Kristin Fritsche"
output: 
 html_document:
    toc: true
    toc_depth: 3
    toc_float: false
    number_sections: true
    code_folding: hide
    
pdf_document:
    fig_caption: true
date: "`r format(Sys.time(), '%d %B, %Y %H:%M:%S')`"
---


```{r, warning=FALSE, message=FALSE}
library(ChAMP)
library(ChAMPdata)
library(pheatmap)
library(dendextend)
#library("IlluminaHumanMethylationEPICanno.ilm10b2.hg19")
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)

library(readxl)


result_folder = paste("./Results",format(Sys.time(), "%Y-%m-%d"),sep="/")
if (!file.exists(result_folder)) dir.create(result_folder, recursive=T)

data_storage_folder = file.path("./Results/data_storage","2021-06-10")
if (!file.exists(data_storage_folder)) dir.create(data_storage_folder, recursive=T)

rerun_preprocessing = F

```

# Experimental Design:

Gastric biopsies from patients with intestinal metaplasia were taken from metaplastic and non-metaplastic control sites. Primary gastric epithelial cells were extracted and expanded on feeder cells. Single cell clones were derived, expanded on feeders and then cultured for few passages as organoid cultures to remove feeder cells. 

DNA was isolated and hybridized to the Illumina EPIC methylation array.

To analyse the generated data in R, we will use and further adapt the "Chip Analysis Methylation Pipeline" (ChAMP) package [Morris J, *et al.*, 2013].


# Samples

```{r}
d = read.table("../../raw_data_array/2020-186-ILL_METMIB_N=9/IDAT/sample_sheet.csv", header=T, sep=",", row.names=NULL)

d[,c("Sentrix_ID","Sentrix_Position","Sample_Name", "Sample_Group","CloneID","PatientID")]
```



# Data Import

Raw intensity values (red and green channel .idat files) are loaded and transformed as beta values. In this process, some filtering steps are already included:
Probes   
- with a detection p-value > 0.01  
- with a bead-count < 3  
- in which the probed CpG falls near a SNP as defined in Zhou W. *et al.*, 2016  
- in which the probe aligns to multiple locations with bwa as defined in Nordlund *et al.*, 2013   
and   
- non-cg probes  
are removed.


```{r, message=TRUE}

set.seed(123456789)

if(rerun_preprocessing){
  
  myLoad = champ.load(directory = "../../raw_data_array/2020-186-ILL_METMIB_N=9/IDAT", arraytype = "EPIC", force = T, method = "minfi")


  save(myLoad, file = file.path(data_storage_folder, "myLoad.RData"))
} else {
  load(file.path(data_storage_folder, "myLoad.RData"))
}


myLoad$pd$Sample_Label = with(myLoad$pd, paste0(PatientID,"_",Sample_Group,"_",CloneID))

```

# Quality Control raw data

The QC function uses the raw beta values for initial quality control:  
- The Multidimensional Scaling Plot (MDS) visualizes the similarity of the samples based on the    top 1,000 most variable probes of all samples.  
- The density Plot displays the beta distribution of each sample  
- The dendrogram clusters all samples (the distance between the samples is calculated directly by   all probes)  


```{r QC raw}

champ.QC()


```


# Normalization

The normlization method applied here is "SWAN (subset-quantile within array normalization)" [Maksimovic J, *et al.*, 2012], which adjusts for the type I and type II bias of the methylation array. 

```{r Normalization} 


if(rerun_preprocessing){
  myNorm = champ.norm(method = "SWAN", arraytype = "EPIC")

  save(myNorm, file = file.path(data_storage_folder,"myNorm.RData"))
} else {
  load(file.path(data_storage_folder, "myNorm.RData"))
}

```

# Quality Control normalized Data  

## MDS plot  
The Multidimensional Scaling Plot (MDS) visualizes the similarity of the samples based on the top 1,000 most variable probes of all samples.


```{r}
## modify mdsPlot function to not show the legend

my_mdsPlot = function (dat, numPositions = 1000, sampNames = NULL, sampGroups = NULL, 
          xlim, ylim, pch = 1, pal = brewer.pal(8, "Dark2"), legendPos = "bottomleft", 
          legendNCol, main = NULL) 
{
  if (is(dat, "MethylSet") || is(dat, "RGChannelSet")) {
    b <- getBeta(dat)
    if (is.null(main)) 
      main <- sprintf("Beta MDS\n%d most variable positions", 
                      numPositions)
  }
  else if (is(dat, "matrix")) {
    b <- dat
    if (is.null(main)) 
      main <- sprintf("MDS\n%d most variable positions", 
                      numPositions)
  }
  else {
    stop("dat must be an 'MethylSet' or 'matrix'.")
  }
  o <- order(-matrixStats::rowVars(b))[1:numPositions]
  d <- dist(t(b[o, ]))
  fit <- cmdscale(d)
  if (missing(xlim)) 
    xlim <- range(fit[, 1]) * 1.2
  if (missing(ylim)) 
    ylim <- range(fit[, 2]) * 1.2
  if (is.null(sampGroups)) 
    sampGroups <- rep(1, numPositions)
  sampGroups <- as.factor(sampGroups)
  col <- pal[sampGroups]
  if (is.null(sampNames)) {
    plot(fit[, 1], fit[, 2], col = col, pch = pch, xlim = xlim, 
         ylim = ylim, xlab = "First dimension", ylab = "Second dimension", main = main, cex.axis = 0.8, cex.lab = 0.8)
  }
  else {
    plot(0, 0, type = "n", xlim = xlim, ylim = ylim, xlab = "First dimension", 
         ylab = "Second dimension", main = main, cex.axis = 0.8, cex.lab = 0.8)
    text(fit[, 1], fit[, 2], sampNames, col = col)
  }
}

```


```{r MDS plot, fig.width=6, fig.height=5}

mdsPlot(myNorm, numPositions = 1000, sampGroups = myLoad$pd$Sample_Group, sampNames = myLoad$pd$Sample_Name, legendPos = "bottomleft")

mdsPlot(myNorm, numPositions = 1000, sampGroups = myLoad$pd$Sample_Group, sampNames = myLoad$pd$PatientID, legendPos = "bottomleft")

#tiff(filename = "MDS.tiff", width = 3, height =3, units = "in", pointsize = 12, compression = c("none"), bg = "transparent", res = 600, family = "")
#my_mdsPlot(myNorm, numPositions = 1000, sampGroups =sample_annotation$Sample_Group, legendPos = "topleft", legendNCol = 1, pch = 16, pal = c("#00afff", "#000fff", "#00d700", "#005f00",  "#ffd700", "#d7af00"), main = "")
#dev.off()

```


## Density Plot  
The density Plot displays the beta distribution of each sample.

```{r density plot}

densityPlot(dat = myNorm, sampGroups = myLoad$pd$Sample_Group, main = paste("Density plot of normalized data (", nrow(myNorm), " probes)", sep = ""))

densityPlot(dat = myNorm, sampGroups = myLoad$pd$PatientID, main = paste("Density plot of normalized data (", nrow(myNorm), " probes)", sep = ""))
```

## Dendrogram  
The dendrogram clusters all samples (the distance between the samples is calculated directly by all probes)

```{r dendrogram}

hc <- hclust(dist(t(myNorm)))
dend <- as.dendrogram(hc)
MyColor <- rainbow(length(table(myLoad$pd$Sample_Group)))
names(MyColor) <- names(table(myLoad$pd$Sample_Group))
labels_colors(dend) <- MyColor[myLoad$pd$Sample_Group[order.dendrogram(dend)]]
dend <- dend %>% set("labels_cex", 0.7)
dend <- dend %>% set("leaves_pch", 19) %>% set("leaves_cex", 0.6) %>% set("leaves_col",         MyColor[myLoad$pd$Sample_Group[order.dendrogram(dend)]])
plot(dend, center = TRUE, main = paste("All samples (", nrow(myNorm), " probes)", sep = ""))
legend("topright", fill = MyColor, legend = names(MyColor))


hc <- hclust(dist(t(myNorm)))
dend <- as.dendrogram(hc)
MyColor <- rainbow(length(table(myLoad$pd$PatientID)))
names(MyColor) <- names(table(myLoad$pd$PatientID))
labels_colors(dend) <- MyColor[myLoad$pd$PatientID[order.dendrogram(dend)]]
dend <- dend %>% set("labels_cex", 0.7)
dend <- dend %>% set("leaves_pch", 19) %>% set("leaves_cex", 0.6) %>% set("leaves_col",         MyColor[as.character(myLoad$pd$PatientID[order.dendrogram(dend)])])
plot(dend, center = TRUE, main = paste("All samples (", nrow(myNorm), " probes)", sep = ""))
legend("topright", fill = MyColor, legend = names(MyColor))

```


## Heatmap  
The Heatmap visualizes the similarity of the samples based on the top 1,000 most variable probes of all samples.

```{r heatmap 1000 most var}
o <- order(-matrixStats::rowVars(myNorm))[1:1000]
data <- myNorm[o,]
col_anno <- as.data.frame(myLoad$pd)[,c("Sample_Group", "PatientID")]
pheatmap(data, cluster_rows = T, main = "Heatmap of 1,000 most variable CpGs", show_rownames = FALSE, annotation_col = col_anno, fontsize_col = 8, labels_col = myLoad$pd[colnames(data),"Sample_Label"])
```

# Batch effect correction  

With the ChAMP SVD function (runs Singular Value Decomposition on a dataset to estimate the impact of batch effects) I test for putative batch effects and perform subsequently batch effect correction using the champ.runCombat method for batch adjustment, adjusting for the slide, which at the same time adjusts for the patient effect as well, because each patient was hybridized to a different slide.

The first plot shows the result of the SVD analysis before adjustment. Then the data are adjusted for "slide". And the second plots shows the result of the SVD analysis after adjustment.   

```{r, message=FALSE}

champ.SVD(RGEffect = TRUE, beta = myNorm)
```

```{r}
data("probe.features.epic")


if(rerun_preprocessing){
  
  myNorm_anno = merge.data.frame(myNorm, probe.features, by=0, sort = F, all.x = T)

  myNorm_anno_sorted = myNorm_anno[order(myNorm_anno$CHR, myNorm_anno$MAPINFO),]
  rownames(myNorm_anno_sorted) = myNorm_anno_sorted$Row.names
  myNorm_anno_sorted$Row.names = NULL

  save(myNorm_anno_sorted, file = file.path(data_storage_folder,"myNorm_anno_sorted.RData")  )
  
}else{
  load(file.path(data_storage_folder, "myNorm_anno_sorted.RData"))
}


myNorm_2 <- merge.data.frame(myNorm, probe.features[,c("CHR", "MAPINFO", "gene", "feature", "cgi", "feat.cgi")], by=0, sort=F, all.x = T)
 
write.table(myNorm_2, file = file.path(result_folder, "Normalized_betas.txt"), row.names=F , sep="\t", dec=".")
save(myNorm_2, file = file.path(data_storage_folder,"Normalized_betas.RData"))

full_anno = getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19, "Other")

full_anno = full_anno[, c("UCSC_RefGene_Name", "UCSC_RefGene_Accession", "UCSC_RefGene_Group", "Phantom5_Enhancers", "DMR", "X450k_Enhancer", "Regulatory_Feature_Group", "GencodeCompV12_NAME", "GencodeCompV12_Accession", "GencodeCompV12_Group", "DNase_Hypersensitivity_NAME", "DNase_Hypersensitivity_Evidence_Count", "OpenChromatin_NAME", "OpenChromatin_Evidence_Count", "TFBS_NAME", "TFBS_Evidence_Count")]

myNorm_2 = merge.data.frame(myNorm_2, full_anno, by=0, sort=F, all.x=T)

write.table(myNorm_2, file = file.path(result_folder, "Normalized_betas_fully_annotated.txt"), row.names = F, sep = "\t", dec = ".")
```

# Conclusion

# Software versions

```{r}
sessionInfo()
```

